### ISSUE TITLE
swift NPU版训练速度与mindspeed训练速度对比问题

### ISSUE THREAD

**Issue by ziliangok** [modelscope/ms-swift#8127]
### Checklist / 检查清单

- [x] I have searched existing issues, and this is a new question or discussion topic. / 我已经搜索过现有的 issues，确认这是一个新的问题与讨论。

### Question Description / 问题描述

直接使用swift + megatron + mindspeed训练，理论上速度应该和mindspeed + megatron速度是相同的，但是我自己做实验，发现swift速度明显比mindspeed快，这是为什么呢？
swift 使用3.12.5版本，mindspeed和megatron使用配套版本。

# swift训练命令
NPROC_PER_NODE=16 megatron sft     --ignore_args_error true     --model /model/Qwen3-8B     --save_safetensors true     --dataset '/datasets/AI-ModelScope/alpaca-gpt4-data-en#500'     --train_type full     --tensor_model_parallel_size 2     --pipeline_model_parallel_size 2     --sequence_parallel true     --micro_batch_size 2     --global_batch_size 512     --finetune true     --cross_entropy_loss_fusion true     --lr 1e-4     --lr_warmup_fraction 0.05     --min_lr 1e-5     --max_length 4096     --no_save_optim true     --no_save_rng true     --dataset_num_proc 4     --log_throughput --log_interval 1     --use_rotary_position_embeddings     --train_iters 10000

step time：16876.9ms, tps 7766.355196

# mindspeed训练命令
torchrun --nproc_per_node 16 --nnodes 1 --node_rank 0 --master_addr localhost --master_port 6000 pretrain_gpt.py --use-mcore-models --spec my_scripts.qwen3_spec layer_spec --qk-layernorm --tokenizer-name-or-path /datasets/deepseek3_hf_tokenizer --max-position-embeddings 4096 --num-layers 36 --hidden-size 4096 --ffn-hidden-size 12288 --num-attention-heads 32 --tokenizer-type PretrainedFromHF --make-vocab-size-divisible-by 1 --rotary-base 1000000 --untie-embeddings-and-output-weights --disable-bias-linear --position-embedding-type rope --normalization RMSNorm --swiglu --attention-softmax-in-fp32 --group-query-attention --num-query-groups 8 --norm-epsilon 1e-6 --data-path /datasets/deepseek3_apalaca_eod/apalca_text_eod_document --split 100,0,0 --log-interval 1 --save-interval 100 --eval-interval 100 --eval-iters 0 --no-load-optim --no-load-rng --use-flash-attn --use-distributed-optimizer --micro-batch-size 2 --global-batch-size 512 --lr 1.25e-6 --lr-decay-style cosine --min-lr 1.25e-7 --weight-decay 1e-1 --lr-warmup-fraction 0.01 --attention-dropout 0.0 --init-method-std 0.01 --hidden-dropout 0.0 --clip-grad 1.0 --adam-beta1 0.9 --adam-beta2 0.95 --initial-loss-scale 4096 --seed 42 --bf16 --train-iters 100 --seq-length 4096 --tensor-model-parallel-size 2 --pipeline-model-parallel-size 2 --attention-backend flash --calculate-per-token-loss --cross-entropy-loss-fusion --expert-tensor-parallel-size 1 --finetune --fp8-amax-compute-algo max --fp8-amax-history-len 1024 --log-memory-to-tensorboard --log-throughput --log-timers-to-tensorboard --log-validation-ppl-to-tensorboard --make-vocab-size-divisible-by 128 --moe-grouped-gemm --moe-router-dtype fp32 --moe-token-dispatcher-type alltoall --num-workers 4 --recompute-granularity selective --recompute-modules core_attn --sequence-parallel --distributed-backend nccl | tee logs/train_mcore_qwen3_8b.log

step time: 40939.7ms, tps 3201.586724

两个框架mindspeed打印的args我也用文本对比工具对比过了，除了swift新增的一些参数，其他基本一致。因为megatron不支持sft，所以我mindspeed实验是用pretrain_gpt跑的，因为没有qwen3的layer-spec，我从mindspeed-llm那里把spec复制过来。

# 通过打印训练的数据，发现没有padding，应该是这个导致的。
